File: implementations/julia-square-matrix/TSPBenchmark/src/espprc-index.jl
using DataStructures

mutable struct LabelWithIndex{ T <: Real }
    at::Int
    visits::Vector{Bool}
    ignore::Bool
    pred::Union{Nothing,Int}
    cost::Float64
    length::T
    q::Vector{Int}
    successors::Vector{Int}
end

function dominates(l1::T, l2::T) where T <: LabelWithIndex
    if l1.cost > l2.cost || l1.length > l2.length
        return false
    end
    for (v1, v2) ∈ zip(l1.visits, l2.visits)
        if v1 > v2
            return false
        end
    end
    for (r1, r2) in zip(l1.q, l2.q)
        if r1 > r2
            return false
        end
    end
    true
end

mutable struct LabelCollection{ T <: Real }
    labels::Vector{LabelWithIndex{T}}
end

# create empty label
function LabelWithIndex(d::MT, nresources::Int) where MT <: TSPData
    LabelWithIndex(1, zeros(Bool, d.n), false, nothing, 0.0, 0,
                   zeros(Int, nresources), zeros(Int, 0))
end

# create empty label and add it to lc
function emptylabel(d::DT, lc::LCT, nresources::Int) where DT <: TSPData where LCT <: LabelCollection
    push!(lc.labels, LabelWithIndex(d, nresources))
    return length(lc.labels)
end

# extend label and add it to LC
function extendlabel(d::MT, lc::LC, from::Int, to::Int) where MT <: TSPData where LC <: LabelCollection
    visits = copy(lc.labels[from].visits)
    visits[to] = true
    q = [ (((to-1) & 1 << (i-1)) > 0) ? r + 1 : r
          for (i,r) in enumerate(lc.labels[from].q) ]
    nl = LabelWithIndex(to, visits, false, from,
                        lc.labels[from].cost + getaux(d, lc.labels[from].at,to),
                        lc.labels[from].length + dist(d, lc.labels[from].at,to),
                        q,
                        zeros(Int, 0))
    push!(lc.labels, nl)
    return length(lc.labels)
end

function marksuccessors!(lc::LC, index::Int) where LC <: LabelCollection
    let l = lc.labels[index]
        for s ∈ l.successors
            lc.labels[s].ignore = true
            marksuccessors!(lc, s)
        end
    end
end

# update given list of labels with new label
# returns true if newlabel is added, false otherwise
function updatedominance!(lc::LCT, labels::Vector{Int}, newlabel::Int) where LCT <: LabelCollection
    i = 1
    while i ≤ length(labels)
        if dominates(lc.labels[labels[i]], lc.labels[newlabel])
            return false
        end
        if dominates(lc.labels[newlabel], lc.labels[labels[i]])
            marksuccessors!(lc, labels[i])
            if i < length(labels)
                labels[i] = labels[end]
            end
            pop!(labels)
        else
            i += 1
        end
    end
    # at this point we know that newlabel is not dominated, so we add it
    push!(labels, newlabel)
    true
end

function solvewithindex(e::ESPPRC{MT, T}) where MT <: TSPData where T <: Real
    # we store all labels here
    lc = LabelCollection(LabelWithIndex{T}[])
    # initial label
    l = emptylabel(e.d, lc, e.nresources)
    Q = Deque{Int}()
    inQ = zeros(Bool, e.d.n)
    push!(Q, 1)
    inQ[1] = true
    labels = [ Vector{Int}() for i in 1:e.d.n ]
    push!(labels[1], l)
    # main DP loop
    while ! isempty(Q)
        n = popfirst!(Q)
        inQ[n] = false
        for lindex ∈ labels[n]
            lc.labels[lindex].ignore && continue
            for succ ∈ 1:e.d.n
                (lc.labels[lindex].visits[succ] || succ == n) && continue
                # is it length-feasible?
                if lc.labels[lindex].length + dist(e.d, n, succ) + dist(e.d, succ, 1) > e.maxlen
                    continue
                end
                # is it resource-feasible?
                rfeas = true
                for (r, q) ∈ enumerate(lc.labels[lindex].q)
                    if ((succ-1) & (1 << (r-1))) > 0 &&
                        q + 1 > e.resourcecapacity
                        rfeas = false
                        break
                    end
                end
                rfeas || continue
                # at this point we know the extension is feasible so we do it
                nl = extendlabel(e.d, lc, lindex, succ)
                added = updatedominance!(lc, labels[succ], nl)
                added && push!(lc.labels[lindex].successors, nl)
                if added && (! inQ[succ]) && succ ≠ 1
                    push!(Q, succ)
                    inQ[succ] = true
                end
            end
            lc.labels[lindex].ignore = true
        end
    end
    Int(trunc(minimum(lc.labels[lindex].cost for lindex in labels[1])))
end

File: implementations/numpy/espprcindex.py
import collections

import numpy as np

import espprc

class LabelWithIndex(espprc.Label):
    # construct an initial label corresponding to an empty path
    def __init__(self):
        # vertex where this partial path is arriving at
        self.at = 0
        # already visited nodes
        self.visited = set()
        # has this label already been extended?
        self.ignore = False
        # predecessor; should point to a Label
        self.pred = -1
        self.cost = 0
        self.length = 0
        # resource consumption
        self.q = np.array([0 for x in self.resources])
        # successors for later recursive deletion when dominated
        self.successors = []

    # Extend label to given vertex
    def extend(self, rc, d, predindex, vertex):
        nl = LabelWithIndex()
        nl.at = vertex
        nl.pred = predindex
        nl.visited = self.visited.union((vertex,))
        nl.cost = self.cost + rc[self.at,vertex]
        nl.length = self.length + d[self.at,vertex]
        # resource is consumed if i^th bit of vertex is 1
        nl.q = np.array([ r + 1 if (vertex & (1 << i) > 0) else r
                          for i, r in enumerate(self.q) ])
        return nl

    def __repr__(self):
        return 'Label at node ' + str(self.at) \
            + '\tcost = ' + str(self.cost) + '\tlength = ' + str(self.length) \
            + '\tq = ' + str(self.q)
    
class LabelCollection:
    def __init__(self):
        self.labels = []
        
    def marksuccessors(self, lindex):
        for s in self.labels[lindex].successors:
            self.labels[s].ignore = True
            self.marksuccessors(s)

    # * compare newlabel with collection of labels removing dominated labels as
    #   necessary
    # * adds newlabel to collection labels if it is not dominated
    # * returns true if newlabel was added
    def updatedominance(self, labels, nlindex):
        i = 0
        while i < len(labels):
            if self.labels[labels[i]].dominates(self.labels[nlindex]):
                return False
            elif self.labels[nlindex].dominates(self.labels[labels[i]]):
                self.marksuccessors(labels[i])
                if i < len(labels) - 1:
                    labels[i] = labels[-1]
                del labels[-1]
            else:
                i += 1
        # at this point no label has dominated newlabel so we add it
        labels.append(nlindex)
        return True

    def emptylabel(self):
        self.labels.append(LabelWithIndex())
        return len(self.labels) - 1

    # extend label at index 'fro' to node 'to'
    def extend(self, rc, d, fro, to):
        self.labels.append(self.labels[fro].extend(rc, d, fro, to))
        return len(self.labels) - 1
    
class ESPPRCLC(espprc.ESPPRC):
    def solve(self):
        # initialise DP data: list of labels, queue, initial label, resources
        vertices = np.array([ x for x in range(self.n) ])
        Q, Qset = collections.deque(), set([0])
        Q.append(0)
        labels = [ [] for x in vertices ]
        LabelWithIndex.resources = np.array([x for x in range(self.nresources)])
        # label collection
        lc = LabelCollection()
        l0 = lc.emptylabel()
        labels[0].append(l0)
        # step 3: run DP
        while len(Qset) > 0:
            n = Q.popleft()
            Qset.remove(n)
            for lindex in labels[n]:
                label = lc.labels[lindex]
                if label.ignore:
                    continue
                for succ in vertices:
                    if succ in label.visited or succ == n: continue
                    # succ is a candidate for extension; is it length-feasible?
                    if label.length + self.d[n,succ] + self.d[succ,0] \
                       > self.maxlen:
                        continue
                    # is it resource-feasible?
                    rfeas = True
                    for i, r in enumerate(label.q):
                        if (succ & (1 << i) > 0) and r + 1 > self.rescap:
                            rfeas = False
                            break
                    if not rfeas: continue
                    # now we can actually extend the label!
                    nl = lc.extend(self.rc, self.d, lindex, succ)
                    # Let's update dominance
                    added = lc.updatedominance(labels[succ], nl)
                    if added:
                        label.successors.append(nl)
                    if added and not (succ in Qset) and succ != 0:
                        Qset.add(succ)
                        Q.append(succ)
                label.ignore = True
        # step 4: return distance of cheapest label as hash value.
        return min(lc.labels[l].cost for l in labels[0])
